platform/mellanox/mlnx-platform-api/sonic_platform/modules_mgmt.py

```diff
@@ -0,0 +1,756 @@
+                    return STATE_HW_PRESENT
+            except Exception as e:
+                logger.log_info("exception {} for port {} setting final state STATE_ERROR_HANDLER".format(e, port))
+                module_sm_obj.set_final_state(STATE_ERROR_HANDLER)
+                return STATE_ERROR_HANDLER
+        module_sm_obj.set_final_state(STATE_HW_NOT_PRESENT)
+        return STATE_HW_NOT_PRESENT
+
+    def check_if_module_available(self, port, module_sm_obj, dynamic=False):
+        logger.log_info("enter check_if_module_available port {} module_sm_obj {}".format(port, module_sm_obj))
+        module_fd_indep_path = SYSFS_INDEPENDENT_FD_POWER_GOOD.format(port)
+        if os.path.isfile(module_fd_indep_path):
+            try:
+                # not using utils.read_int_from_file since need to catch the exception here if no such file or it is
+                # not accesible. utils.read_int_from_file will return 0 in such a case
+                module_power_good_fd = open(module_fd_indep_path, "r")
+                val = module_power_good_fd.read()
+                val_int = int(val)
+                module_sm_obj.module_power_good_fd_path = module_fd_indep_path
+                module_sm_obj.module_power_good_fd = module_power_good_fd
+
+                if 0 == val_int:
+                    logger.log_info(f'port {port} power is not good')
+                    module_sm_obj.set_final_state(STATE_HW_NOT_PRESENT)
+                    return STATE_HW_NOT_PRESENT
+                elif 1 == val_int:
+                    logger.log_info(f'port {port} power is good')
+                    return STATE_MODULE_AVAILABLE
+            except Exception as e:
+                logger.log_info("exception {} for port {}".format(e, port))
+                return STATE_HW_NOT_PRESENT
+        logger.log_info(f'port {port} has no power good file {module_fd_indep_path}')
+        module_sm_obj.set_final_state(STATE_HW_NOT_PRESENT)
+        return STATE_HW_NOT_PRESENT
+
+    def check_if_power_on(self, port, module_sm_obj, dynamic=False):
+        logger.log_info(f'enter check_if_power_on for port {port}')
+        module_fd_indep_path = SYSFS_INDEPENDENT_FD_POWER_ON.format(port)
+        if os.path.isfile(module_fd_indep_path):
+            try:
+                val_int = utils.read_int_from_file(module_fd_indep_path)
+                if 0 == val_int:
+                    logger.log_info(f'check_if_power_on port {port} is not powered')
+                    return STATE_NOT_POWERED
+                elif 1 == val_int:
+                    logger.log_info(f'check_if_power_on port {port} is powered')
+                    return STATE_POWERED
+            except Exception as e:
+                logger.log_info(f'check_if_power_on got exception {e}')
+                module_sm_obj.set_final_state(STATE_HW_NOT_PRESENT)
+                return STATE_HW_NOT_PRESENT
+
+    def power_on_module(self, port, module_sm_obj, dynamic=False):
```

Comment:
@dbarashinvd where is this called?

Reply:
all of these state machine functions are called from get_sm_func, which takes each time the next state to run (the next function).
you can see the list of function and its resolution in the get_sm_func function.
it's called both in the static detection and in the dynamic detection since basically it's the same flow on both, the flow that is run to detect the modules properly.
from a check that the cable is plugged in, to the power on check, and through power good check, power cap check, module type check and so on.
until final decision if it's FW control or SW control module.